client/opentracing/src/main/java/org/hawkular/apm/client/opentracing/TraceContext.java

```diff
@@ -61,25 +61,33 @@
 
     private TraceRecorder recorder;
     private ContextSampler sampler;
+    private Property serviceName;
+    private Property buildTimeStamp;
 
     private static List<NodeProcessor> nodeProcessors = new ArrayList<>();
 
     static {
         nodeProcessors.add(new DefaultNodeProcessor());
     }
 
+    private static Property ENV_SERVICE_NAME = new Property(Constants.PROP_SERVICE_NAME, getServiceName());
+    private static Property ENV_BUILD_STAMP = new Property(Constants.PROP_BUILD_STAMP, getBuildStamp());
+
     /**
      * This constructor initialises the trace context.
      *
      * @param topSpan The top level span in the trace
      * @param rootNode The builder for the root node of the trace fragment
      * @param recorder The trace recorder
+     * @param serviceName The service name to be assigned to the root node, or null if a service name should be guessed
      */
-    public TraceContext(APMSpan topSpan, NodeBuilder rootNode, TraceRecorder recorder, ContextSampler sampler) {
+    public TraceContext(APMSpan topSpan, NodeBuilder rootNode, TraceRecorder recorder, ContextSampler sampler, String serviceName, String buildStamp) {
         this.topSpan = topSpan;
         this.rootNode = rootNode;
         this.recorder = recorder;
         this.sampler = sampler;
+        this.serviceName = null == serviceName ? ENV_SERVICE_NAME : new Property(Constants.PROP_SERVICE_NAME, serviceName);
+        this.buildTimeStamp = null == buildStamp ? ENV_BUILD_STAMP : new Property(Constants.PROP_BUILD_STAMP, buildStamp);
```

Comment:
We would cast the `TraceRecorder` to `EnvironmentAwareTraceRecorder`, if possible.